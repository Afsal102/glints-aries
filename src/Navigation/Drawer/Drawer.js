import React, { Component } from 'react';

import {
  DrawerContainer,
  DrawerWrapper,
} from '../../Style/Navigation/DrawerStyle';

class Drawer extends Component<State, Props> {
  constructor() {
    super();
    this.state = {
      isDisplay: false,
    };
  }

  handleStart = () => {
    this.setState({ isDisplay: true });
  }

  handleAnimationEnd = () => {
    this.setState({ isDisplay: false });
  }

  componentDidUpdate() {
    const { isDisplay } = this.state;
    const { isOpen } = this.props;

    if (isOpen && !isDisplay) {
      this.handleStart();
    }
  }

  render() {
    const {
      children,
      isOpen,
      onClose,
      ...defaultProps
    } = this.props;

    const {
      isDisplay,
    } = this.state;

    return (
      <DrawerContainer
        aria-modal="true"
        aria-hidden={isDisplay ? 'false' : 'true'}
        isDisplay={isDisplay}
        open={isOpen}
        onAnimationStart={this.handleAnimationStart}
        onAnimationEnd={this.handleAnimationEnd}
        onClick={() => onClose()}
      >
        <DrawerWrapper
          role="dialog"
          open={isOpen}
          tabIndex="0"
          onClick={e => e.stopPropagation()}
          {...defaultProps}
        >
          {children}
        </DrawerWrapper>
      </DrawerContainer>
    );
  }
}
type Props = {
  children: React$Node,
  isOpen: boolean,
  onClose: Function,
}

type State = {
  isDisplay: boolean
}

export default Drawer;
