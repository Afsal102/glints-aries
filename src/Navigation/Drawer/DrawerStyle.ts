import styled, { keyframes } from 'styled-components';
import { DrawerPosition } from './Drawer';

const openFromRight = keyframes`
  from {
    transform: translateX(100%);
  }

  to {
    transform: translateX(0%);
  }
`;

const closeToRight = keyframes`
  from {
    transform: translateX(0%);
  }

  to {
    transform: translateX(100%);
  }
`;

const openFromLeft = keyframes`
  from {
    transform: translateX(-100%);
  }

  to {
    transform: translateX(0%);
  }
`;

const closeToLeft = keyframes`
  from {
    transform: translateX(0%);
  }

  to {
    transform: translateX(-100%);
  }
`;

const fadeIn = keyframes`
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
`;

const fadeOut = keyframes`
  from {
    opacity: 1;
  }

  to {
    opacity: 0;
  }
`;

export const DrawerContainer = styled.div<DrawerContainerProps>`
  position: fixed;
  display: ${({ isDisplay }) => (isDisplay ? 'block' : 'none')};
  min-height: 100vh;
  max-height: 100vh;
  width: 100vw;
  top: 0px;
  left: 0px;
  background-color: rgba(0, 0, 0, 0.5);
  animation: ${({ open }) => (open ? fadeIn : fadeOut)} 0.3s ease-in-out;
  z-index: ${({ isDisplay }) => (isDisplay ? '9999' : '-1')};
`;

interface DrawerContainerProps {
  isDisplay: boolean;
  open: boolean;
}

export const DrawerWrapper = styled.div<DrawerWrapperProps>`
  position: absolute;
  height: 100vh;
  width: 80vw;
  max-width: 300px;
  top: 0px;
  ${({ position }) =>
    ({
      left: `
    left: 0px;
  `,
      right: `
    right: 0px;
  `,
    }[position])}

  background-color: white;
  animation: ${({ open, position }) =>
      ({
        right: open ? openFromRight : closeToRight,
        left: open ? openFromLeft : closeToLeft,
      }[position])}
    0.3s ease-in-out;
  transform: ${({ open, position }) =>
    ({
      right: open ? 'translateX(0%)' : 'translateX(100%)',
      left: open ? 'translateX(0%)' : 'translateX(-100%)',
    }[position])};
  outline: none;
  box-shadow: -2px 0 8px rgba(0, 0, 0, 0.15);
  overflow-y: auto;
`;

interface DrawerWrapperProps {
  open: boolean;
  position: DrawerPosition;
}
