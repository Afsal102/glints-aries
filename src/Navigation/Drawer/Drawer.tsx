import * as React from 'react';

import { DrawerContainer, DrawerWrapper } from './DrawerStyle';

const Drawer = ({ children, isOpen, onClose, ...defaultProps }: Props) => {
  const [isDisplay, setIsDisplay] = React.useState<boolean>(false);

  const handleAnimationStart = () => {
    setIsDisplay(true);
  };

  const handleAnimationEnd = () => {
    setIsDisplay(false);
  };

  React.useEffect(
    function startAnimationOnOpen() {
      if (isOpen && !isDisplay) {
        handleAnimationStart();
      }
    },
    [isDisplay, isOpen]
  );

  return (
    <DrawerContainer
      className="aries-drawer"
      data-testid="drawer-container"
      aria-modal="true"
      aria-hidden={isDisplay ? 'false' : 'true'}
      isDisplay={isDisplay}
      open={isOpen}
      onAnimationStart={handleAnimationStart}
      onAnimationEnd={handleAnimationEnd}
      onClick={() => onClose()}
    >
      <DrawerWrapper
        className="drawer-contentwrapper"
        role="dialog"
        data-testid="drawer-wrapper"
        open={isOpen}
        tabIndex={0}
        onClick={e => e.stopPropagation()}
        {...defaultProps}
      >
        {children}
      </DrawerWrapper>
    </DrawerContainer>
  );
};

interface Props {
  children: React.ReactNode;
  isOpen: boolean;
  onClose: Function;
}

export default Drawer;
