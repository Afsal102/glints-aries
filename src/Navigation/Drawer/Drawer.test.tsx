import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render, queryByTestId } from '@testing-library/react';

import Drawer, { DrawerPosition } from './Drawer';

describe('<Drawer /> render', () => {
  const matchSnapshot = (position: DrawerPosition) => {
    test(`should match snapshot when position is ${position} and drawer is open`, () => {
      const snapshot = renderer
        .create(
          <Drawer isOpen={true} position={position} onClose={jest.fn()}>
            test-content
          </Drawer>
        )
        .toJSON();
      expect(snapshot).toMatchSnapshot();
    });

    test(`should match snapshot when position is ${position} and drawer is closed`, () => {
      const snapshot = renderer
        .create(
          <Drawer isOpen={false} position={position} onClose={jest.fn()}>
            test-content
          </Drawer>
        )
        .toJSON();
      expect(snapshot).toMatchSnapshot();
    });
  };

  ['left', 'right', undefined].forEach(matchSnapshot);
});

describe('<Drawer/> behaviour', () => {
  test('onClose callback should be called when container is clicked', async () => {
    const onClose = jest.fn();
    render(
      <Drawer isOpen={false} onClose={onClose}>
        test-content
      </Drawer>
    );

    const drawerWrapper = queryByTestId(document.body, 'drawer-wrapper');
    fireEvent.click(drawerWrapper);

    expect(onClose).not.toHaveBeenCalled();

    const drawerContainer = queryByTestId(document.body, 'drawer-container');
    fireEvent.click(drawerContainer);

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should remove overflow hidden from document body after close', () => {
    const onClose = jest.fn();

    const { rerender } = render(
      <Drawer isOpen={false} onClose={onClose}>
        test-content
      </Drawer>
    );
    const drawerContainer = queryByTestId(document.body, 'drawer-container');
    expect(drawerContainer).toHaveAttribute('aria-hidden', 'true');

    // rerender here will render the same component again, thus triggering the
    // internal update logic
    rerender(
      <Drawer isOpen={true} onClose={onClose}>
        test-content
      </Drawer>
    );
    expect(drawerContainer).toHaveAttribute('aria-hidden', 'false');
  });
});
