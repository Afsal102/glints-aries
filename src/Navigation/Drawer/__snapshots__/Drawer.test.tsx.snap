// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> render should match snapshot when position is left and drawer is closed 1`] = `
<div
  aria-hidden="true"
  aria-modal="true"
  className="DrawerStyle__DrawerContainer-sc-13skgii-0 wwLow aries-drawer"
  data-testid="drawer-container"
  onClick={[Function]}
  open={false}
>
  <div
    className="DrawerStyle__DrawerWrapper-sc-13skgii-1 dNGthb drawer-contentwrapper"
    data-testid="drawer-wrapper"
    onClick={[Function]}
    open={false}
    role="dialog"
    tabIndex={0}
  >
    test-content
  </div>
</div>
`;

exports[`<Drawer /> render should match snapshot when position is left and drawer is open 1`] = `
<div
  aria-hidden="false"
  aria-modal="true"
  className="DrawerStyle__DrawerContainer-sc-13skgii-0 fYRGbP aries-drawer"
  data-testid="drawer-container"
  onClick={[Function]}
  open={true}
>
  <div
    className="DrawerStyle__DrawerWrapper-sc-13skgii-1 fPmHMH drawer-contentwrapper"
    data-testid="drawer-wrapper"
    onClick={[Function]}
    open={true}
    role="dialog"
    tabIndex={0}
  >
    test-content
  </div>
</div>
`;

exports[`<Drawer /> render should match snapshot when position is right and drawer is closed 1`] = `
<div
  aria-hidden="true"
  aria-modal="true"
  className="DrawerStyle__DrawerContainer-sc-13skgii-0 wwLow aries-drawer"
  data-testid="drawer-container"
  onClick={[Function]}
  open={false}
>
  <div
    className="DrawerStyle__DrawerWrapper-sc-13skgii-1 enMoix drawer-contentwrapper"
    data-testid="drawer-wrapper"
    onClick={[Function]}
    open={false}
    role="dialog"
    tabIndex={0}
  >
    test-content
  </div>
</div>
`;

exports[`<Drawer /> render should match snapshot when position is right and drawer is open 1`] = `
<div
  aria-hidden="false"
  aria-modal="true"
  className="DrawerStyle__DrawerContainer-sc-13skgii-0 fYRGbP aries-drawer"
  data-testid="drawer-container"
  onClick={[Function]}
  open={true}
>
  <div
    className="DrawerStyle__DrawerWrapper-sc-13skgii-1 PuCTW drawer-contentwrapper"
    data-testid="drawer-wrapper"
    onClick={[Function]}
    open={true}
    role="dialog"
    tabIndex={0}
  >
    test-content
  </div>
</div>
`;

exports[`<Drawer /> render should match snapshot when position is undefined and drawer is closed 1`] = `
<div
  aria-hidden="true"
  aria-modal="true"
  className="DrawerStyle__DrawerContainer-sc-13skgii-0 wwLow aries-drawer"
  data-testid="drawer-container"
  onClick={[Function]}
  open={false}
>
  <div
    className="DrawerStyle__DrawerWrapper-sc-13skgii-1 enMoix drawer-contentwrapper"
    data-testid="drawer-wrapper"
    onClick={[Function]}
    open={false}
    role="dialog"
    tabIndex={0}
  >
    test-content
  </div>
</div>
`;

exports[`<Drawer /> render should match snapshot when position is undefined and drawer is open 1`] = `
<div
  aria-hidden="false"
  aria-modal="true"
  className="DrawerStyle__DrawerContainer-sc-13skgii-0 fYRGbP aries-drawer"
  data-testid="drawer-container"
  onClick={[Function]}
  open={true}
>
  <div
    className="DrawerStyle__DrawerWrapper-sc-13skgii-1 PuCTW drawer-contentwrapper"
    data-testid="drawer-wrapper"
    onClick={[Function]}
    open={true}
    role="dialog"
    tabIndex={0}
  >
    test-content
  </div>
</div>
`;
